utils/pattern/src/interpolator/mod.rs

```diff
@@ -0,0 +1,302 @@
+/// fallible. As such, the decision has been made to design the API in line with what
+/// we hope will become a trait signature of a fallible iterator in the future, rather
+/// than implementing a reversed infallible iterator (where the [`Item`] would be
+/// `Option<Result<Item>>`).
+///
+/// Since the interpolator chains on top of the [`Parser`] it inherits the same fallible
+/// iterator API and behavior.
+///
+/// [`Item`]: std::iter::Iterator::Item
+/// [`HashMap`]: std::collections::HashMap
+/// [`Parser`]: crate::parser::Parser
+/// [`IntoIterVec`]: crate::pattern::IntoIterVec
+pub struct Interpolator<P, R, K, E>
+where
+    R: ReplacementProvider<K, E>,
+{
+    pattern: P,
+    replacements: R,
+    current_replacement: Option<R::Iter>,
+}
+
+impl<P, R, K, E> Interpolator<P, R, K, E>
+where
+    R: ReplacementProvider<K, E>,
+{
+    /// Creates a new `Interpolator`.
+    ///
+    /// # Examples
+    /// ```
+    /// use icu_pattern::{Parser, Interpolator};
+    ///
+    /// enum Element {
+    ///     Literal(String),
+    ///     Token,
+    /// }
+    ///
+    /// let mut parser = Parser::new("{0}, {1}", false);
+    /// let mut interpolator = Interpolator::<_, _, _, Element>::new(parser, vec![
+    ///     vec![
+    ///         Element::Token
+    ///     ]
+    /// ]);
+    /// ```
+    pub fn new(pattern: P, replacements: R) -> Self {
+        Self {
+            pattern,
+            replacements,
+            current_replacement: None,
+        }
+    }
+
+    /// An iterator method that advances the iterator and returns the result of an attempt to
+    /// interpolate parser and replacement provider tokens.
+    ///
+    /// # Examples
+    /// ```
+    /// use icu_pattern::{Parser, Interpolator};
+    ///
+    /// #[derive(Debug, PartialEq)]
+    /// enum Element {
+    ///     Literal(String),
+    ///     TokenOne,
+    ///     TokenTwo,
+    /// }
+    ///
+    /// impl From<&str> for Element {
+    ///     fn from(input: &str) -> Self {
+    ///        Self::Literal(input.to_string())
+    ///     }
+    /// }
+    ///
+    /// let mut parser = Parser::new("{0}, {1}", false);
+    /// let mut interpolator = Interpolator::new(parser, vec![
+    ///     vec![
+    ///         Element::TokenOne
+    ///     ],
+    ///     vec![
+    ///         Element::TokenTwo
+    ///     ]
+    /// ]);
+    ///
+    ///
+    /// // A call to try_next() returns the next value…
+    /// assert_eq!(Ok(Some(Element::TokenOne)), interpolator.try_next());
+    /// assert_eq!(Ok(Some(Element::Literal(", ".to_string()))), interpolator.try_next());
+    /// assert_eq!(Ok(Some(Element::TokenTwo)), interpolator.try_next());
+    ///
+    /// // … and then None once it's over.
+    /// assert_eq!(Ok(None), interpolator.try_next());
+    /// ```
+    ///
+    /// # Lifetimes
+    ///
+    /// - `p`: The life time of an input parser, which is the life time of the string slice to be parsed.
+    pub fn try_next<'p>(&mut self) -> Result<E, K>
```

Comment:
Shouldn't we just return InterpolateError?

Reply:
Instead of a `Result`? I don't think I understand.